File: src/SWA.Ariadne.Logic/SolverFactory.cs
using System;
using System.Collections.Generic;
using System.Text;
using SWA.Ariadne.Model;
using SWA.Ariadne.Settings;

namespace SWA.Ariadne.Logic
{
    /// <summary>
    /// Knows the various MazeSolver strategies and can create instances of them.
    /// </summary>
    public static class SolverFactory
    {
        /// <summary>
        /// Gets an array with the Types of all implemented MazeSolver strategies.
        /// </summary>
        public static System.Type[] SolverTypes
        {
            get { return solverTypes; }
        }
        private static System.Type[] solverTypes = new System.Type[] {
            typeof(OpposedBacktracker),
            typeof(ProximityBacktracker),
            typeof(RandomBacktracker),
            typeof(MasterSolver),
            typeof(RightHandWalker),
            typeof(LeftHandWalker),
            typeof(RandomWalker),
            typeof(RoundRobinFlooder),
            typeof(CloseFlooder),
            typeof(FarFlooder),
            typeof(OpposedFlooder),
            typeof(ProximityFlooder),
            typeof(HesitatingFlooder),
            typeof(CenterFlooder),
            typeof(CornerFlooder),
            typeof(ForwardFlooder),
            typeof(BackwardFlooder),
            typeof(ThickestBranchFlooder),
            typeof(ThinnestBranchFlooder),
            typeof(SpreadingFlooder),
            typeof(RandomFlooder),
        };

        /// <summary>
        /// Returns true if a DeadEndChecker may be installed in the given solverType.
        /// </summary>
        /// <param name="solverType"></param>
        /// <returns></returns>
        public static bool HasEfficientVariant(System.Type solverType)
        {
            bool result = (solverType.IsSubclassOf(typeof(SolverBase)));
            foreach (System.Type t in noEfficientSolverTypes)
            {
                result &= (solverType != t);
            }
            return result;
        }
        private static System.Type[] noEfficientSolverTypes = new System.Type[] {
            typeof(MasterSolver),
            typeof(RandomWalker),
        };
        public static string EfficientPrefix = "Efficient";

        private static Type SolverType(string name)
        {
            foreach (Type t in SolverTypes)
            {
                if (t.Name == name)
                {
                    return t;
                }
            }
            return null;
        }

        /// <summary>
        /// Returns a new MazeSolver of the given Type.
        /// </summary>
        /// <param name="solverType"></param>
        /// <param name="maze"></param>
        /// <returns></returns>
        private static IMazeSolver CreateSolver(Type solverType, Maze maze, IMazeDrawer mazeDrawer)
        {
            IMazeSolver result = (IMazeSolver)solverType.GetConstructor(
                new Type[2] { typeof(Maze), typeof(IMazeDrawer) }).Invoke(
                new object[2] { maze, mazeDrawer }
                );
            result.Reset();

            return result;
        }

        /// <summary>
        /// Returns a new MazeSolver.
        /// A (reasonably) intelligent strategy is chosen randomly.
        /// </summary>
        /// <param name="maze"></param>
        /// <returns></returns>
        private static IMazeSolver CreateSolver(Maze maze, IMazeDrawer mazeDrawer)
        {
            Random r = SWA.Utilities.RandomFactory.CreateRandom();

            while (true)
            {
                Type t = solverTypes[r.Next(solverTypes.Length)];
                bool shouldBeEfficient = (r.Next(2) == 0);
                shouldBeEfficient &= RegisteredOptions.GetBoolSetting(RegisteredOptions.OPT_EFFICIENT_SOLVERS);

                if (t == typeof(RandomWalker))
                {
                    // too dumb
                    continue;
                }

                if (t == typeof(MasterSolver))
                {
                    // too smart
                    continue;
                }

                IMazeSolver result = CreateSolver(t, maze, mazeDrawer);

                if (shouldBeEfficient && HasEfficientVariant(t))
                {
                    result.MakeEfficient();
                }

                return result;
            }
        }

        /// <summary>
        /// Returns a new MazeSolver.
        /// If the strategyName is valid, that type is created; otherwise a random type is returned.
        /// </summary>
        public static IMazeSolver CreateSolver(string strategyName, Maze maze, IMazeDrawer mazeDrawer)
        {
            IMazeSolver result;
            bool isEfficient = false;

            if (strategyName != null)
            {
                if (strategyName.StartsWith(EfficientPrefix))
                {
                    strategyName = strategyName.Substring(EfficientPrefix.Length);
                    isEfficient = true;
                }
            }

            Type strategy = SolverType(strategyName);
            if (strategy != null)
            {
                // If strategyName is a valid solver type name:
                result = CreateSolver(strategy, maze, mazeDrawer);
                if (isEfficient)
                {
                    result.MakeEfficient();
                }
            }
            else
            {
                // Otherwise (strategy name is "any"):
                result = CreateSolver(maze, mazeDrawer);
            }

            return result;
        }

        /// <summary>
        /// The default MazeSolver strategy.
        /// </summary>
        public static Type DefaultStrategy
        {
            get { return typeof(RandomBacktracker); }
        }

        /// <summary>
        /// Returns a new MazeSolver.
        /// The default strategy.
        /// </summary>
        /// <param name="maze"></param>
        /// <returns></returns>
        public static IMazeSolver CreateDefaultSolver(Maze maze, IMazeDrawer mazeDrawer)
        {
            return CreateSolver(DefaultStrategy, maze, mazeDrawer);
        }

        /// <summary>
        /// Returns the list of squares on the solution path from start to end square.
        /// </summary>
        /// <param name="maze"></param>
        /// <returns></returns>
        public static List<MazeSquare> SolutionPath(Maze maze)
        {
            // Note: This code must not modify the given maze.
            MasterSolver m = new MasterSolver(maze, null);
            return m.Path;
        }
    }
}

File: src.JAVA/src/swa/ariadne/logic/factory/SolverFactory.java
package swa.ariadne.logic.factory;

import java.lang.reflect.Type;
import java.util.List;

import swa.ariadne.logic.IMazeDrawer;
import swa.ariadne.logic.IMazeSolver;
import swa.ariadne.logic.MasterSolver;
import swa.ariadne.logic.RandomBacktracker;
import swa.ariadne.logic.RoundRobinFlooder;
import swa.ariadne.model.Maze;
import swa.ariadne.model.MazeSquare;
import swa.util.RandomFactory;

/**
 * Knows the various MazeSolver strategies and can create instances of them.
 * 
 * @author Stephan.Wacker@web.de
 */
public final
class SolverFactory
{
    //--------------------- Static Member variables and Properties

    /** The Types of all implemented MazeSolver strategies. */
    private static Type[] solverTypes = {
        // TODO:
        // ProximityBacktracker.class,
        // OpposedBacktracker.class,
        RandomBacktracker.class,
        MasterSolver.class,
        // RightHandWalker.class,
        // LeftHandWalker.class,
        // RandomWalker.class,
        RoundRobinFlooder.class,
        // CloseFlooder.class,
        // FarFlooder.class,
        // OpposedFlooder.class,
        // ProximityFlooder.class,
        // HesitatingFlooder.class,
        // CenterFlooder.class,
        // CornerFlooder.class,
        // ForwardFlooder.class,
        // BackwardFlooder.class,
        // RandomForwardFlooder.class,
        // RandomBackwardFlooder.class,
        // ThickestBranchFlooder.class,
        // ThinnestBranchFlooder.class,
        // SpreadingFlooder.class,
        // RandomFlooder.class,
    };
    
    /**
     *  A collection of all available IMazeSolver constructors.
     *  TODO: use solvers instead of solverTypes
     */
    private static SolverConstructor[] solvers = {
        Solvers.RandomBacktracker(),
        Solvers.MasterSolver(),
        Solvers.RoundRobinFlooder(),
    };

    /**
     * @return An array with all implemented MazeSolver strategies.
     */
    public static Type[] getSolverTypes()
    {
        return solverTypes;
    }

    /**
     * The default MazeSolver strategy.
     */
    public static final String DefaultStrategy = "RandomBacktracker";

    /**
     *  A prefix that may be added to a strategy name.
     *  If the strategyName passed to CreateSolver() begins with this prefix,
     *  the returned MazeSolver will employ a dead-end detection algorithm.
     *  @see #createSolver(String, Maze, IMazeDrawer)
     */
    public static final String EfficientPrefix = "Efficient";

    //--------------------- Static methods for creating a MazeSolver

    /**
     * @param solverType A specific MazeSolver type. 
     * @param maze The problem to be solved.
     * @param mazeDrawer An object that can draw the maze. 
     * @return A new MazeSolver of the given Type.
     */
    private static IMazeSolver createSolver(Type solverType, Maze maze, IMazeDrawer mazeDrawer)
    {
        IMazeSolver result;
        
        try
        {
            result = (IMazeSolver) solverType.getClass().getConstructor(Maze.class, IMazeDrawer.class).newInstance(maze, mazeDrawer);
            result.reset();
        }
        catch (Exception e)
        {
            result = createDefaultSolver(maze, mazeDrawer);
        }

        return result;
    }

    /**
     * If the given strategyName is valid, that type is created; otherwise a random type is returned.
     * @param strategyName Name of the Solver Type.
     * @param maze The problem to be solved.
     * @param mazeDrawer An object that can draw the maze. 
     * @return A new MazeSolver of the given type name.
     */
    public static IMazeSolver createSolver(String strategyName, Maze maze, IMazeDrawer mazeDrawer)
    {
        IMazeSolver result;
        boolean isEfficient = false;

        /* TODO
        if (strategyName != null)
        {
            if (strategyName.StartsWith(EfficientPrefix))
            {
                strategyName = strategyName.Substring(EfficientPrefix.Length);
                isEfficient = true;
            }
        }
        */

        Type strategy = getSolverType(strategyName);
        if (strategy != null)
        {
            // If strategyName is a valid solver type name:
            result = createSolver(strategy, maze, mazeDrawer);
            if (isEfficient)
            {
                // TODO result.MakeEfficient();
            }
        }
        else
        {
            // Otherwise (strategy name is "any"):
            result = createSolver(maze, mazeDrawer);
        }

        return result;
    }

    /**
     * @param maze The problem to be solved.
     * @param mazeDrawer An object that can draw the maze. 
     * @return A new MazeSolver of a random type.
     */
    private static IMazeSolver createSolver(Maze maze, IMazeDrawer mazeDrawer)
    {
        while (true)
        {
            Type t = solverTypes[RandomFactory.nextInt(solverTypes.length)];
            boolean shouldBeEfficient = (RandomFactory.nextInt(2) == 0);
            // TODO: shouldBeEfficient &= RegisteredOptions.GetBoolSetting(RegisteredOptions.OPT_EFFICIENT_SOLVERS);

            /* TODO
            if (t == RandomWalker.class)
            {
                // too dumb
                continue;
            }
            */

            if (t == MasterSolver.class)
            {
                // too smart
                continue;
            }

            IMazeSolver result = createSolver(t, maze, mazeDrawer);

            /* TODO:
            if (shouldBeEfficient && HasEfficientVariant(t))
            {
                result.MakeEfficient();
            }
            */

            return result;
        }
    }
    
    /**
     * @param maze The problem to be solved.
     * @param mazeDrawer An object that can draw the maze. 
     * @return A MazeSolver object implementing the default strategy.
     */
    public static IMazeSolver createDefaultSolver(Maze maze, IMazeDrawer mazeDrawer)
    {
        IMazeSolver result = new RandomBacktracker(maze, mazeDrawer);
        result.reset();
        return result;
    }

    //--------------------- Static methods related to the solution

    /**
     * @param maze The problem to be solved.
     * @return The list of squares on the solution path from start to target square.
     */
    public static List<MazeSquare> getSolutionPath(Maze maze)
    {
        // Note: This code must not modify the given maze.
        MasterSolver m = new MasterSolver(maze, null);
        return m.getPath();
    }

    //--------------------- Auxiliary methods

    /**
     * @param name The name of one of the solver types. 
     * @return A valid MazeSolver Type or null if the name is not declared.
     */
    private static Type getSolverType(String name)
    {
        for (Type t : solverTypes)
        {
            if (t.toString().equals(name))
            {
                return t;
            }
        }
        return null;
    }
}
